storage-operator/pkg/controller/volume/volume_controller.go

```diff
@@ -24,149 +26,761 @@ import (
+		if value != VOLUME_PROTECTION {
+			newFinalizers = append(newFinalizers, value)
+		}
+	}
+
+	volume.SetFinalizers(newFinalizers)
+	return self.client.Update(context.TODO(), volume)
+}
+
+func (self *ReconcileVolume) removePvFinalizer(
+	pv *corev1.PersistentVolume,
+) error {
+	finalizers := pv.GetFinalizers()
+
+	newFinalizers := finalizers[:0]
+	for _, value := range finalizers {
+		if value != VOLUME_PROTECTION {
+			newFinalizers = append(newFinalizers, value)
+		}
+	}
+
+	pv.SetFinalizers(newFinalizers)
+	return self.client.Update(context.TODO(), pv)
+}
+
+// Reconcile reads that state of the cluster for a Volume object and makes
+// changes based on the state read and what is in the Volume.Spec
+//
 // Note:
-// The Controller will requeue the Request to be processed again if the returned error is non-nil or
-// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
-func (r *ReconcileVolume) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+// The Controller will requeue the Request to be processed again if the returned
+// error is non-nil or Result.Requeue is true, otherwise upon completion it
+// will remove the work from the queue.
+func (r *ReconcileVolume) Reconcile(
+	request reconcile.Request,
+) (reconcile.Result, error) {
 	reqLogger := log.WithValues("Request.Name", request.Name)
-	reqLogger.Info("Reconciling Volume")
+	reqLogger.Info("reconciling volume: BEGIN")
+	defer reqLogger.Info("reconciling volume: STOP")
 
-	// Fetch the Volume instance
-	instance := &storagev1alpha1.Volume{}
-	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	volume := &storagev1alpha1.Volume{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, volume)
 	if err != nil {
 		if errors.IsNotFound(err) {
-			// Request object not found, could have been deleted after reconcile request.
-			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
-			// Return and don't requeue
-			// TODO Remove the finalizer from the generated PV, if exists
+			// Volume not found:
+			// => all the finalizers have been removed & Volume has been deleted
+			// => there is nothing left to do
+			reqLogger.Info("volume already deleted: nothing to do")
 			return reconcile.Result{}, nil
 		}
-		// Error reading the object - requeue the request.
+		reqLogger.Error(err, "cannot read Volume: requeue")
 		return reconcile.Result{}, err
 	}
 
-	result, err := r.salt.TestPing()
-	if err != nil {
-		reqLogger.Error(err, "test.ping failed")
-		return reconcile.Result{}, err
+	// Check if the volume is marked for deletion (i.e., deletion tsamp is set).
+	if volume.GetDeletionTimestamp() != nil &&
+		volume.Status.Phase != storagev1alpha1.VolumeTerminating {
+		if volume.Status.Phase == storagev1alpha1.VolumePending {
+			reqLogger.Info("pending volume cannot be marked for deletion: requeue")
+			return reconcile.Result{Requeue: true}, nil
```

Comment:
Whenever requeueing, consider whether a delay should be set, otherwise you can end up with a spinning/busy loop.

Reply:
From what I've seen in my testing, Kubernetes/SDK/something seems to detect when an "error" is repeated and adds a delay by itself (I've seen that when I reschedule during Salt job polling).

But couldn't hurt to have an explicit delay: should we have something constant or increasing?

Reply:
Done (1s delay, constant)